### PR TITLE
#231 rebased on main for PyTorch 2.5.1

### PR DESCRIPTION
**Pull request by Tobias-Fischer**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Very hacky for now without proper commit messages etc; just for testing

### CODE REVIEW

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
FYI @jeongseok-meta @baszalmstra - I created this PR for playing around; I'll see if I have time to push this further. Do we have a test server that we can play with easily? Unfortunately the conda-forge Windows machine only has 20GB storage left, which won't be nearly enough for this ..

**Comment by baszalmstra** [comment]
I see you already found how to get access to the windows runners! 👍

**Comment by Tobias-Fischer** [comment]
Hmm I’m not sure why the CI isn’t running

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
@baszalmstra @hmaarrfk - it seems like it compiled libtorch just fine but then failed with compiling the python part for py313. I’m not sure if it’s an issue with the sccache or the CMakeCache rewrite to replace the Python version. I’m on mobile only today - if you have any ideas please let me know

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please rerender

**Comment by baszalmstra** [comment]
I ran into the same issue in my PR. The only reliable way I found to fix this was to add `cmake --build build --target clean` after each output build and rely on sccache for caching instead (see my PR for details)… Far from ideal..

**Comment by Tobias-Fischer** [comment]
Hmm this directly builds on top of your PR and already includes the clean: https://github.com/Tobias-Fischer/pytorch-cpu-feedstock/blob/4bbb512e60f2ddb643c1a57c174df94106abf957/recipe/bld.bat#L139

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -16,9 +16,11 @@ blas_impl:
...
 # https://github.com/conda-forge/.cirun
-github_actions_labels:          # [linux]
+github_actions_labels:          # [linux or win]
 - cirun-openstack-gpu-2xlarge   # [linux]
+- cirun-azure-windows-2xlarge   # [win]
 
 megabuild:
 - true      # [linux]
 - false     # [osx]
+- true      # [win]
```
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:26]: I would avoid megabuild and focus on the abseil challenges

File: recipe/meta.yaml [hunk 2]
```diff
@@ -50,15 +52,18 @@ build:
   ignore_run_exports:
     - python *                               # [megabuild]
     - numpy *                                # [megabuild]
-  skip: true  # [win]
+
+  # TODO(baszalmstra): Remove this before merging
+  skip: true  # [not win or not (cuda_compiler_version == "12.0" or cuda_compiler == "None")]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:57]:
```suggestion
  skip: true  # [not win or not (cuda_compiler_version == "12.0" or cuda_compiler == "None")]
  skip: true  # [py!=312]
```

just to save some resources on future builds. definitely let this one complete.

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
mkl builds fail with:
```
2024-12-15T21:49:51.9936405Z import: 'torch'
2024-12-15T21:49:52.0741907Z Traceback (most recent call last):
2024-12-15T21:49:52.0742455Z   File "C:\bld\libtorch_1734290771319\test_tmp\run_test.py", line 2, in <module>
2024-12-15T21:49:52.0744309Z     import torch
2024-12-15T21:49:52.0744746Z   File "C:\bld\libtorch_1734290771319\_test_env\lib\site-packages\torch\__init__.py", line 262, in <module>
2024-12-15T21:49:52.0745261Z     _load_dll_libraries()
2024-12-15T21:49:52.0745855Z   File "C:\bld\libtorch_1734290771319\_test_env\lib\site-packages\torch\__init__.py", line 245, in _load_dll_libraries
2024-12-15T21:49:52.0746632Z     raise err
2024-12-15T21:49:52.0747344Z OSError: [WinError 127] The specified procedure could not be found. Error loading "C:\bld\libtorch_1734290771319\_test_env\lib\site-packages\torch\lib\fbgemm.dll" or one of its dependencies.
2024-12-15T21:49:52.5360364Z WARNING: Tests failed for pytorch-2.5.1-cpu_mkl_py310h279fb04_106.conda - moving package to C:\bld\broken
2024-12-15T21:49:52.6222842Z Traceback (most recent call last):
2024-12-15T21:49:52.6223286Z   File "D:\Miniforge\Lib\site-packages\conda_build\build.py", line 3481, in test
2024-12-15T21:49:52.6244074Z     utils.check_call_env(
2024-12-15T21:49:52.6244474Z   File "D:\Miniforge\Lib\site-packages\conda_build\utils.py", line 406, in check_call_env
2024-12-15T21:49:52.6254024Z     return _func_defaulting_env_to_os_environ("call", *popenargs, **kwargs)
2024-12-15T21:49:52.6255833Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-12-15T21:49:52.6256364Z   File "D:\Miniforge\Lib\site-packages\conda_build\utils.py", line 382, in _func_defaulting_env_to_os_environ
2024-12-15T21:49:52.6257743Z     raise subprocess.CalledProcessError(proc.returncode, _args)
2024-12-15T21:49:52.6258738Z subprocess.CalledProcessError: Command '['C:\\Windows\\system32\\cmd.exe', '/d', '/c', 'C:\\bld\\libtorch_1734290771319\\test_tmp\\conda_test_runner.bat']' returned non-zero exit status 1.
```

non-mkl builds fail with:
```
2024-12-15T21:49:55.5505504Z import: 'torch'
2024-12-15T21:49:56.1187098Z WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
2024-12-15T21:49:56.1187940Z E0000 00:00:1734299396.118431    2620 descriptor_database.cc:633] File already exists in database: onnx/onnx_onnx_torch-ml.proto
2024-12-15T21:49:56.1189034Z F0000 00:00:1734299396.118613    2620 descriptor.cc:2236] Check failed: GeneratedDatabase()->Add(encoded_file_descriptor, size) 
```

Will try and fix the non-mkl builds by setting `ONNX_USE_PROTOBUF_SHARED_LIBS=ON`. Not sure about the mkl builds.

**Comment by Tobias-Fischer** [comment]
I have a feeling that the mkl issue is due to https://github.com/conda-forge/conda-forge.github.io/issues/1597 but I don't understand enough of it. 

In the logs we can see that
```
2024-12-15T21:48:30.6185793Z ClobberWarning: This transaction has incompatible packages due to a shared path.
2024-12-15T21:48:30.6186765Z   packages: conda-forge/win-64::intel-openmp-2024.2.1-h57928b3_1083, conda-forge/win-64::openmp-5.0.0-vc14_1
2024-12-15T21:48:30.6187424Z   path: 'library/bin/libiomp5md.dll'
```

We also know the mkl error is:
```
2024-12-15T21:49:52.0747344Z OSError: [WinError 127] The specified procedure could not be found. Error loading "C:\bld\libtorch_1734290771319\_test_env\lib\site-packages\torch\lib\fbgemm.dll" or one of its dependencies.
```

And that one of the fbgemm.dll dependencies is libiomp5md.dll (`(libtorch,Lib/site-packages/torch/lib/fbgemm.dll): Needed DSO Library/bin/libiomp5md.dll found in ['conda-forge/win-64::intel-openmp==2024.2.1=h57928b3_1083'])`.

Not sure how to go from here, @isuruf mentioned in https://github.com/conda-forge/conda-forge.github.io/issues/1597 that:
> > Of course**. But isn't the underlying issue here that conda-forge's mkl [depends](https://github.com/conda-forge/intel_repack-feedstock/blob/master/recipe/meta.yaml#L91-L106) on intel-openmp?
> 
> That's because of pytorch. If we have a pytorch windows build, then we can make mkl depend on llvm-openmp. pytorch upstream seems to be linking to the vcomp interface in libiomp5md.dll which is not present in llvm-openmp, but we can avoid that if we have a build on windows.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -234,12 +267,14 @@ outputs:
...
         - libgomp   # [linux]
         - llvm-openmp    # [osx]
+        # TODO(baszalmstra): There are linker errors on Windows when using protobuf from conda-forge
         - libabseil
         - libprotobuf
         - sleef
         - libuv
         - pkg-config  # [unix]
         - typing_extensions
+        - onnx 1.16  # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:277]: I really don't think this works.... basically from what I can tell, many people vendor onnx and just assume that this is the way life goes.

https://github.com/conda-forge/onnx-feedstock/pull/113

is my attempt at making a library out of it. but i don't think i ever managed to finish it.

Comment (hmaarrfk) [reply recipe/meta.yaml:277]: onnx for windows should be compiled in the same way as onnx for linux is compiled.

It seems the errors were abseil related. is abseil coming from the windows side?

Comment (Tobias-Fischer) [reply recipe/meta.yaml:277]: I don't think it's an abseil issue, is it? All that abseil provides is a warning that messages are written to stderr before initialising. The real problem seems to stem from protobuf, which states that "descriptor_database.cc:633] File already exists in database: onnx/onnx_onnx_torch-ml.proto"

There are some mentions in onnx issues: https://github.com/search?q=repo%3Aonnx%2Fonnx+%22File+already+exists+in+database%22&type=issues

**Comment by Tobias-Fischer** [comment]
It seems like USE_LITE_PROTO fixed both mkl and generic builds :).

What are the best steps to get a proper review going? We need:
- general cleanup, removing of old comments etc
- CUDA builds
- Proper rebuild with commit history
- ?

I’ll kick off some CUDA builds - let’s see how we go.

**Comment by Tobias-Fischer** [comment]
Closing - all work has been rebased on top of #231

**Pull request closed without merging**